File: shot-consumer/app/src/main/java/com/karumi/screenshot/ui/view/SuperHeroDetailActivity.java
/*
 * Copyright (C) 2017 Karumi.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.karumi.screenshot.ui.view;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;
import com.karumi.screenshot.R;
import com.karumi.screenshot.SuperHeroesApplication;
import com.karumi.screenshot.model.SuperHero;
import com.karumi.screenshot.ui.presenter.SuperHeroDetailPresenter;
import com.squareup.picasso.Picasso;
import javax.inject.Inject;
import butterknife.Bind;

public class SuperHeroDetailActivity extends BaseActivity implements SuperHeroDetailPresenter.View {

  private static final String SUPER_HERO_NAME_KEY = "super_hero_name_key";

  @Inject SuperHeroDetailPresenter presenter;

  @Bind(R.id.iv_super_hero_photo) ImageView superHeroPhotoImageView;
  @Bind(R.id.tv_super_hero_name) TextView superHeroNameTextView;
  @Bind(R.id.tv_super_hero_description) TextView superHeroDescriptionTextView;
  @Bind(R.id.iv_avengers_badge) View avengersBadgeView;

  @Override protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    initializeDagger();
    initializePresenter();
  }

  @Override public int getLayoutId() {
    return R.layout.super_hero_detail_activity;
  }

  @Override public void showSuperHero(SuperHero superHero) {
    Picasso.with(this).load(superHero.getPhoto()).fit().centerCrop().into(superHeroPhotoImageView);
    superHeroNameTextView.setText(superHero.getName());
    superHeroDescriptionTextView.setText(superHero.getDescription());
    int avengersBadgeVisibility = superHero.isAvenger() ? View.VISIBLE : View.GONE;
    avengersBadgeView.setVisibility(avengersBadgeVisibility);
  }

  public static void open(Context context, String superHeroName) {
    Intent intent = new Intent(context, SuperHeroDetailActivity.class);
    intent.putExtra(SUPER_HERO_NAME_KEY, superHeroName);
    context.startActivity(intent);
  }

  @Override protected void initializeToolbar() {
    super.initializeToolbar();
    String superHeroName = getSuperHeroName();
    setTitle(superHeroName);
  }

  private void initializeDagger() {
    SuperHeroesApplication app = (SuperHeroesApplication) getApplication();
    app.getMainComponent().inject(this);
  }

  private void initializePresenter() {
    presenter.setView(this);
    String name = getSuperHeroName();
    presenter.setName(name);
    presenter.initialize();
  }

  private String getSuperHeroName() {
    return getIntent().getExtras().getString(SUPER_HERO_NAME_KEY);
  }
}

File: shot-consumer/app2/src/main/java/com/karumi/ui/view/SuperHeroDetailActivity.kt
package com.karumi.ui.view

import android.app.Activity
import android.content.Intent
import androidx.appcompat.widget.Toolbar
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.core.widget.ContentLoadingProgressBar
import com.github.salomonbrys.kodein.Kodein.Module
import com.github.salomonbrys.kodein.bind
import com.github.salomonbrys.kodein.instance
import com.github.salomonbrys.kodein.provider
import com.karumi.R
import com.karumi.domain.model.SuperHero
import com.karumi.domain.usecase.GetSuperHeroByName
import com.karumi.ui.presenter.SuperHeroDetailPresenter
import com.karumi.ui.utils.setImageBackground

class SuperHeroDetailActivity : BaseActivity(), SuperHeroDetailPresenter.View {

    companion object {
        private const val SUPER_HERO_NAME_KEY = "super_hero_name_key"

        fun open(activity: Activity, superHeroName: String) {
            val intent = Intent(activity, SuperHeroDetailActivity::class.java)
            intent.putExtra(SUPER_HERO_NAME_KEY, superHeroName)
            activity.startActivity(intent)
        }
    }

    private val progressBar: ContentLoadingProgressBar by lazy { findViewById(R.id.progress_bar) }
    private val tvSuperHeroDescription: TextView
        by lazy { findViewById(R.id.tv_super_hero_description) }
    private val tvSuperHeroName: TextView
            by lazy { findViewById(R.id.tv_super_hero_name) }
    private val ivSuperHeroPhoto: ImageView
            by lazy { findViewById(R.id.iv_super_hero_photo) }
    private val ivAvengersBadge: ImageView
            by lazy { findViewById(R.id.iv_avengers_badge) }

    override val presenter: SuperHeroDetailPresenter by injector.instance()

    override val layoutId: Int = R.layout.super_hero_detail_activity
    override val toolbarView: Toolbar by lazy { findViewById(R.id.toolbar) }

    override fun preparePresenter(intent: Intent?) {
        val superHeroName = intent?.extras?.getString(SUPER_HERO_NAME_KEY)
        title = superHeroName
        presenter.preparePresenter(superHeroName)
    }

    override fun close() = finish()

    override fun showLoading() {
        progressBar.visibility = View.VISIBLE
    }

    override fun hideLoading() {
        progressBar.visibility = View.GONE
    }

    override fun showSuperHero(superHero: SuperHero) {
        tvSuperHeroName.text = superHero.name
        tvSuperHeroDescription.text = superHero.description
        ivAvengersBadge.visibility =
                if (superHero.isAvenger) View.VISIBLE else View.GONE
        ivSuperHeroPhoto.setImageBackground(superHero.photo)
    }

    override val activityModules = Module(allowSilentOverride = true) {
        bind<SuperHeroDetailPresenter>() with provider {
            SuperHeroDetailPresenter(this@SuperHeroDetailActivity, instance())
        }
        bind<GetSuperHeroByName>() with provider { GetSuperHeroByName(instance()) }
    }
}